packages/contracts/src/utils/JwtRegistry.sol

```diff
@@ -0,0 +1,109 @@
+        dkimRegistry = new DKIMRegistry(address(this));
+    }
+
+    /// @notice Checks if a public key hash is valid and not revoked for a given kis and iss.
+    /// @param domainName The domain name contains kis, iss and azp fields.
+    /// @param publicKeyHash The public key hash to validate.
+    /// @return bool Returns true if the public key hash is valid and not revoked, false otherwise.
+    function isDKIMPublicKeyHashValid(
+        string memory domainName,
+        bytes32 publicKeyHash
+    ) public view returns (bool) {
+        string[] memory parts = this.stringToArray(domainName);
+        string memory kidAndIss = string(abi.encode(parts[0], "|", parts[1]));
+        return dkimRegistry.isDKIMPublicKeyHashValid(kidAndIss, publicKeyHash) 
+          && whitelistedClients[parts[2]];
+    }
+
+    /// @notice Sets a public key hash for a `kis|iss` string  after validating the provided signature.
+    /// @param domainName The domain name contains kis, iss and azp fields.
+    /// @param publicKeyHash The public key hash to set.
+    /// @dev This function requires that the public key hash is not already set or revoked.
+    function setDKIMPublicKeyHash(
+        string memory domainName,
+        bytes32 publicKeyHash
+    ) public {
+        require(bytes(domainName).length != 0, "Invalid domain name");
+        require(publicKeyHash != bytes32(0), "Invalid public key hash");
+        string[] memory parts = this.stringToArray(domainName);
+        string memory kidAndIss = string(abi.encode(parts[0], "|", parts[1]));
+        require(
+            isDKIMPublicKeyHashValid(domainName, publicKeyHash) == false,
+            "publicKeyHash is already set"
+        );
+        require(
+            dkimRegistry.revokedDKIMPublicKeyHashes(publicKeyHash) == false,
+            "publicKeyHash is revoked"
+        );
+
+        dkimRegistry.setDKIMPublicKeyHash(kidAndIss, publicKeyHash);
+        // Register azp
+        whitelistedClients[parts[2]] = true;
+    }
+
+    /// @notice Revokes a public key hash for `kis|iss` string after validating the provided signature.
+    /// @param domainName The domain name contains kis, iss and azp fields.
+    /// @param publicKeyHash The public key hash to revoke.
+    /// @dev This function requires that the public key hash is currently set and not already revoked.
+    function revokeDKIMPublicKeyHash(
+        string memory domainName,
+        bytes32 publicKeyHash
+    ) public {
+        require(bytes(domainName).length != 0, "Invalid domain name");
+        require(publicKeyHash != bytes32(0), "Invalid public key hash");
+        require(
+            isDKIMPublicKeyHashValid(domainName, publicKeyHash) == true,
+            "publicKeyHash is not set"
+        );
+        require(
+            dkimRegistry.revokedDKIMPublicKeyHashes(publicKeyHash) == false,
+            "publicKeyHash is already revoked"
+        );
+
+        dkimRegistry.revokeDKIMPublicKeyHash(publicKeyHash);
+    }
```

Comment:
> Maybe only if the corresponding secret key is leaked which I believe is very unlikely. Would love to hear your thoughts.

We prepare this function for this case.
Although I don't any case where a JWT signer reveals the secret key for the revoked public key, it might happen in the future.